x/proof/keeper/msg_server_submit_proof.go

```diff
@@ -127,12 +224,141 @@ func (k msgServer) queryAndValidateClaimForProof(ctx context.Context, proof *typ
+		)
+	}
+
+	if sessionHeader.GetSessionId() != expectedSessionHeader.GetSessionId() {
+		return types.ErrProofInvalidRelay.Wrapf(
+			"sessionHeaders session IDs mismatch expect: %q, got: %q",
+			expectedSessionHeader.GetSessionId(),
+			sessionHeader.GetSessionId(),
+		)
+	}
+
+	return nil
+}
+
+// verifyClosestProof verifies the closest merkle proof against the expected root hash.
+func verifyClosestProof(
+	proof *smt.SparseMerkleClosestProof,
+	expectedRootHash []byte,
+) error {
+	spec := smt.NoPrehashSpec(sha256.New(), true)
+
+	valid, err := smt.VerifyClosestProof(proof, expectedRootHash, spec)
+	if err != nil {
+		return err
+	}
+
+	if !valid {
+		return types.ErrProofInvalidProof.Wrap("invalid closest merkle proof")
+	}
+
+	return nil
+}
+
+// validateMiningDifficulty ensures that the relay's mining difficulty meets the required
+// difficulty.
+// TODO_TECHDEBT: Factor out the relay mining difficulty validation into a shared function
+// that can be used by both the proof and the miner packages.
+func validateMiningDifficulty(relayBz []byte) error {
+	hasher := sha256.New()
+	hasher.Write(relayBz)
+	realyHash := hasher.Sum(nil)
+
+	difficultyBits, err := protocol.CountDifficultyBits(realyHash)
+	if err != nil {
+		return types.ErrProofInvalidRelay.Wrapf(
+			"error counting difficulty bits: %s",
+			err,
+		)
+	}
+
+	if difficultyBits < relayMinDifficultyBits {
+		return types.ErrProofInvalidRelay.Wrapf(
+			"relay difficulty %d is less than the required difficulty %d",
+			difficultyBits,
+			relayMinDifficultyBits,
+		)
+	}
+
+	return nil
+}
+
+// validateClosestPath ensures that the proof's path matches the expected path.
+func (k msgServer) validateClosestPath(
+	ctx context.Context,
+	proof *smt.SparseMerkleClosestProof,
+	sessionHeader *sessiontypes.SessionHeader,
+) error {
+	// The RelayMiner has to wait until the createClaimWindowStartHeight and the
+	// submitProofWindowStartHeight are open to respectively create the claim and
+	// submit the proof.
+	// These windows are calculated as SessionEndBlockHeight + GracePeriodBlockCount.
+	// see: relayerSessionsManager.waitForEarliest{CreateClaim,SubmitProof}Height().
+	// The RelayMiner hast to wait this long to ensure that late relays are accepted
+	// and included in the SessionNumber=N tree.
+	// (i.e. relays initiated by Applications/Gateways in SessionNumber=N but
+	// arriving to the RelayMiner in SessionNumber=N + 1)
+	// Otherwise, the RelayMiner would not be able to include the late relays in
+	// the SessionNumber N claim and the proof.
+	// Since smt.ProveClosest is in terms of submitProofWindowStartHeight, this
+	// block's hash needs to be used for validation too.
+	// TODO_TECHDEBT(#409): Reference the session rollover documentation here.
+	blockHeight := sessionHeader.GetSessionEndBlockHeight() + sessionkeeper.GetSessionGracePeriodBlockCount()
+	blockHash := k.sessionKeeper.GetBlockHash(ctx, blockHeight)
+
+	if !bytes.Equal(proof.Path, blockHash) {
```

Comment:
⚠️IMPORTANT BLOCKER ⚠️

I don't understand this. Why are we comparing the proof path to the block hash?

This is the sort of thing we really need a unit test for or AT LEAST a TODO_TEST.

Reply:
@Olshansk Here `proof.Path` is the path used in the `ClosestProof` method to obtain the pseudo-random leaf. So if these 2 are different someone is trying to do something dodgy or something has gone wrong.

Reply:
> ⚠️IMPORTANT BLOCKER ⚠️
> 
> I don't understand this. Why are we comparing the proof path to the block hash?
> 
> This is the sort of thing we really need a unit test for or AT LEAST a TODO_TEST.

This case is covered in #417 by https://github.com/pokt-network/poktroll/pull/417/files#diff-9ebcaa7396c928d8fc0a011521652f5f7332d928e25291cf7d04805cf98ec4dcR725.

Reply:
It depends on the hashing function used. A cosmos block hash will be 32 bytes so it presumably uses some variation of `sha256`. As long as the trie is setup with a 256 bit path hasher then yes this is guarenteed to always be the same length (this is the case currently). However, there is no enforcement - a check could be done in the smt's `ClosestProof` logic that throws an error if the hash provided is not the same size as the path hasher of the trie I can add this to https://github.com/pokt-network/smt/pull/42 now

Reply:
> but instead ensure we are using the same sized path hashers (which AFAIK we are)

Exactly. How do we ensure this?

1. **Current**: Manual checking by someone (i.e. @h5law) across two github repositories with implicit institutional knowledge of how things work and relate.
2. Proposed (explicit, commented & clear):
```go
// BlockHashToClosestMerkleProofHashPath converts the block hash,
// a pseudo-random value that cannot be predicted, to a random path
// in the tree.
func BlockHashToClosestMerkleProofHashPath(blockHash string) string {
     smtPathSize = smt.GetPathSize() // this is a new function
     return blockHash[:smtPathSize]
}
```
3. I believe there are two `TODO_BLOCKER`s we need to add here:

```
TODO_BLOCKER: Update `blockHeight` to be the value when the `ProofWindow` opens once the variable is added
TODO_BLOCKER: The seed of the path should be `foo(blockHash, sessionId)` to prevent all proofs needing to use the same path.
```

Reply:
@red-0ne Let's add TODO_BLOCKERs and pass-through functions where appropriate and get this PR in. It'll simplify and unblock a lot.

Reply:
@Olshansk @red-0ne https://github.com/pokt-network/smt/commit/e3dbbbd9f97d4f9007c67f5c7e512a93c8f3335c

Now there is enforcement in ClosestProof and its verification that the paths are the same size in bytes - both methods will error if they dont match which means this case is already covered by verifying the proof - and creating it.

Reply:
Added the `TODO_BLOCKER`s. And now using https://github.com/pokt-network/smt/commit/e3dbbbd9f97d4f9007c67f5c7e512a93c8f3335c.

@Olshansk , Anything else preventing from removing the `TODO_IN_THIS_PR` ?